### PR TITLE
Automatically reset identify

### PR DESCRIPTION
**Pull request by timgl**
## Changes

If app tries to `identify` a user that's already identified (non-anonymous ID), we assume it's an impersonation so we reset.

## Checklist
- [ ] Tests for new code (if applicable)
- [ ] TypeScript definitions (module.d.ts) updated and in sync with library exports (if applicable)

### CODE REVIEW

**Comment by timgl** [comment]
@macobo this is ready for a look now (sorry it took a while!)

File: src/posthog-core.js [hunk 1]
```diff
@@ -798,9 +798,13 @@ PostHogLib.prototype.identify = function (new_distinct_id, userProperties) {
...
     }
 
-    // send an $identify event any time the distinct_id is changing - logic on the server
-    // will determine whether or not to do anything with it.
-    if (new_distinct_id !== previous_distinct_id) {
+    // send an $identify event any time the distinct_id is changing and the old ID is an anoymous ID
+    // - logic on the server will determine whether or not to do anything with it.
+    if (
+        new_distinct_id !== previous_distinct_id &&
+        (!this['persistence']?.props?.$device_id || previous_distinct_id === this['persistence']?.props?.$device_id)
```
Comment (macobo) [inline src/posthog-core.js:805]: `?.` will cause errors on older browsers, we have not yet configured babel to handle it.

I think the right way to handle this is via `this.get_property('$device_id')`

File: src/posthog-core.js [hunk 2]
```diff
@@ -798,9 +798,13 @@ PostHogLib.prototype.identify = function (new_distinct_id, userProperties) {
...
+    ) {
+        console.log(new_distinct_id)
```
Comment (macobo) [inline src/posthog-core.js:807]: Debug logging.

File: src/__tests__/posthog-core.js [hunk 3]
```diff
@@ -36,7 +36,9 @@ describe('identify()', () => {
         expect(given.overrides.register).toHaveBeenCalledWith({ distinct_id: 'a-new-id' })
     })
 
-    it('calls capture when identity changes', () => {
+    it('calls capture when identity changes and old ID is anonymous', () => {
+        given.lib.persistence = { $device_id: 'oldIdentity' }
```
Comment (macobo) [inline src/__tests__/posthog-core.js:40]: Nit: The test is not consistent with how given2 is normally used, causing unneeded repetition between tests.

The way to do this via given2 is following.

Above:
```js
given('overrides', () => ({
  persistence: { $device_id: given.deviceId }
  // same as rest

given('deviceId', () => 'oldIdentity')

// Drop lines like given.lib.persistence = in both tests
// In the second test
given('deviceId', () => 'anonymousId')
```

**Review by macobo — CHANGES_REQUESTED** [review]
Some code-level issues otherwise looks to be doing the right thing. :)

One thing I'm wondering about: What if user is using $app in multiple tabs? What is the expected and behavior after this change?

File: src/__tests__/posthog-core.js (hunk 3 above)
Comment (timgl) [reply src/__tests__/posthog-core.js]: I spent ~30 mins trying to figure out how to do this (in combo with using get_property like above) and couldn't work it out, so i've gone down the previous route again.

Comment (macobo) [reply src/__tests__/posthog-core.js]: Yikes! Sorry about that - happy to help set things up better next time.

Until then, pushed a quick fix myself.

**Comment by macobo** [comment]
Merged. What shall we put to changelog?

**Pull request merged**